criu/mount-v2.c

```diff
@@ -0,0 +1,1159 @@
+				}
+				list_add(&sg->siblings, &p->children);
+				/* Don't break to check for parent collision */
+			}
+
+			/* External slavery */
+			if (!sg->parent && list_empty(&sg->siblings)) {
+				struct mount_info *ext;
+				struct sharing_group *s;
+				char *source = NULL;
+
+				/*
+				 * Though we don't have parent sharing group
+				 * (inaccessible sharing), we can still have
+				 * siblings, so collect them to the list.
+				 */
+				list_for_each_entry(s, &sharing_groups, list) {
+					if (s->master_id != sg->master_id)
+						continue;
+
+					if (s->parent) {
+						pr_err("External slavery sharing group (%d, %d) has parent (%d, %d)\n",
+						       sg->shared_id, sg->master_id, s->parent->shared_id,
+						       s->parent->master_id);
+						return -1;
+					}
+
+					if (!list_empty(&s->siblings)) {
+						pr_err("External slavery sharing group collision (%d, %d) (%d, %d)\n",
+						       sg->shared_id, sg->master_id, s->shared_id, s->master_id);
+						return -1;
+					}
+					list_add(&s->siblings, &sg->siblings);
+				}
+
+				BUG_ON(list_empty(&sg->mnt_list));
+				mi = list_entry(sg->mnt_list.next, struct mount_info, mnt_sharing);
+
+				if ((ext = mnt_get_external_bind_nodev(mi)))
+					source = ext->external;
+				else if (mnt_is_root_bind(mi))
+					source = opts.root;
+
+				if (!source) {
+					pr_err("Sharing group (%d, %d) "
+					       "has unreachable sharing. Try --enable-external-masters.\n",
+					       sg->shared_id, sg->master_id);
+					return -1;
+				}
+
+				sg->source = source;
+				list_for_each_entry(s, &sg->siblings, siblings)
+					s->source = sg->source;
+
+				pr_debug("Detected external slavery for shared group (%d, %d) with source %s\n",
+					 sg->shared_id, sg->master_id, source);
+			}
+		}
+	}
+
+	return 0;
+}
+
+static int propagate_mount_v2(struct mount_info *mi)
+{
+	struct mount_info *t;
+
+	list_for_each_entry(t, &mi->mnt_bind, mnt_bind) {
+		if (t->mounted)
+			continue;
+		if (t->bind)
+			continue;
+		if (!issubpath(t->root, mi->root))
+			continue;
+		pr_debug("\t\tPropagate %d to %d\n", mi->mnt_id, t->mnt_id);
+		t->bind = mi;
+		t->s_dev_rt = mi->s_dev_rt;
+	}
+
+	return 0;
+}
+
+static int do_new_mount_v2(struct mount_info *mi)
+{
+	unsigned long sflags = mi->sb_flags;
+	unsigned long mflags = mi->flags & (~MS_PROPAGATE);
+	char *src;
+	struct fstype *tp = mi->fstype;
+	bool remount_ro = (tp->restore && mi->sb_flags & MS_RDONLY);
+	mount_fn_t do_mount = (tp->mount) ? tp->mount : do_simple_mount;
+
+	src = resolve_source(mi);
+	if (!src)
+		return -1;
+
+	/* Merge superblock and mount flags if it's possible */
+	if (!(mflags & ~MS_MNT_KNOWN_FLAGS) && !((sflags ^ mflags) & MS_RDONLY)) {
+		sflags |= mflags;
+		mflags = 0;
+	}
+
+	if (remount_ro)
+		sflags &= ~MS_RDONLY;
+
+	if (do_mount(mi, src, mnt_fsname(mi), sflags) < 0) {
+		pr_perror("Can't mount at %s", mi->plain_mountpoint);
+		return -1;
+	}
+
+	if (mount(NULL, mi->plain_mountpoint, NULL, MS_PRIVATE, NULL)) {
```

Comment:
Why do we need this extra mount call?  Could you add a comment that explain this?

Reply:
This is a place where we make sure that the mount created by do_new_mount_v2() is 100% private and we would never get any unexpected propagation with it on mount tree construction stage. We do MS_PRIVATE mount call in all places just after mounting something including do_bind_mount_v2(), do_mount_root_v2() and make_yard().

I understand that this call can be considered "excess" as if we know that parent mount is 100% private and we create a fresh new (not bindmount) children mount for it child would be private too, and also even for bindmount of private source new mount would also be private. But I prefer do one extra call for mount syscall but make code simpler so that it is just obvious from the nearby code that newly created mount would be private.

Not sure if I need to add this explanation everywhere or just some note on top of mount-v2.c would be enough?